File: apps/android/app/src/main/java/com/aegispay/android/ui/wallet/WalletViewModel.kt
package com.aegispay.android.ui.wallet

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aegispay.android.network.AegisApiService
import com.aegispay.android.network.Account
import com.aegispay.android.network.FxRateRepository
import com.aegispay.android.network.FxRates
import com.aegispay.android.network.TopUpConfirmRequest
import com.aegispay.android.network.TopUpIntentRequest
import java.math.BigDecimal
import java.math.RoundingMode
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

// ── Constants ─────────────────────────────────────────────────────────────────

/** Hard INR balance ceiling — Stripe KYC threshold. */
val BALANCE_LIMIT_INR: BigDecimal = BigDecimal("100000")

// ── UI State ──────────────────────────────────────────────────────────────────

sealed interface WalletUiState {
    data object Loading : WalletUiState
    data class  Success(
        val accounts:    List<Account>,
        val topUpResult: TopUpResult? = null,
    ) : WalletUiState
    data class  Error(val message: String) : WalletUiState
}

enum class TopUpResult { SUCCESS, FAILED }

// ── ViewModel ─────────────────────────────────────────────────────────────────

@HiltViewModel
class WalletViewModel @Inject constructor(
    private val api:           AegisApiService,
    private val fxRateRepo:    FxRateRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow<WalletUiState>(WalletUiState.Loading)
    val uiState: StateFlow<WalletUiState> = _uiState.asStateFlow()

    /**
     * The Stripe [clientSecret] to present via [PaymentSheet].
     * Set when the backend creates a PaymentIntent; cleared after the sheet is shown.
     */
    private val _clientSecret = MutableStateFlow<String?>(null)
    val clientSecret: StateFlow<String?> = _clientSecret.asStateFlow()

    /**
     * The PaymentIntent ID extracted from the clientSecret (e.g. "pi_xxx" from "pi_xxx_secret_yyy").
     * Passed to [confirmTopUp] after the Stripe SDK reports [PaymentSheetResult.Completed].
     */
    var pendingPaymentIntentId: String? = null
        private set

    private val _isTopUpLoading = MutableStateFlow(false)
    val isTopUpLoading: StateFlow<Boolean> = _isTopUpLoading.asStateFlow()

    // ── FX rates ──────────────────────────────────────────────────────────────

    private val _fxRates   = MutableStateFlow(FxRates())
    val fxRates: StateFlow<FxRates> = _fxRates.asStateFlow()

    private val _fxLoading = MutableStateFlow(false)
    val fxLoading: StateFlow<Boolean> = _fxLoading.asStateFlow()

    init {
        loadAccounts()
        refreshFxRates()
    }

    // ── Load accounts + FX rates ──────────────────────────────────────────────

    fun loadAccounts() {
        viewModelScope.launch {
            _uiState.value = WalletUiState.Loading
            val accountsDeferred = async { runCatching { api.getMyAccount() } }
            val ratesDeferred    = async { fxRateRepo.rates() }

            _fxRates.value = ratesDeferred.await()
            _uiState.value = accountsDeferred.await().fold(
                onSuccess = { WalletUiState.Success(it) },
                onFailure = { WalletUiState.Error(it.message ?: "Failed to load wallet") },
            )
        }
    }

    fun refreshFxRates() {
        viewModelScope.launch {
            _fxLoading.value = true
            _fxRates.value   = fxRateRepo.rates()
            _fxLoading.value = false
        }
    }

    // ── Balance limit helpers ─────────────────────────────────────────────────

    /** INR available balance, or null when no INR account. */
    val inrAvailable: BigDecimal?
        get() = (uiState.value as? WalletUiState.Success)
            ?.accounts?.firstOrNull { it.currency == "INR" }
            ?.availableBalance

    /** Headroom until the INR wallet cap. */
    val inrRoom: BigDecimal
        get() = (BALANCE_LIMIT_INR - (inrAvailable ?: BigDecimal.ZERO))
            .coerceAtLeast(BigDecimal.ZERO)

    /** Returns true when adding [amount] INR would breach the limit. */
    fun wouldExceedLimit(amount: BigDecimal): Boolean =
        (inrAvailable ?: BigDecimal.ZERO) + amount > BALANCE_LIMIT_INR

    // ── Step 1: create PaymentIntent ──────────────────────────────────────────

    /**
     * Calls the backend to create a Stripe PaymentIntent.
     * On success sets [clientSecret] which triggers [WalletScreen] to call
     * `paymentSheet.presentWithPaymentIntent(clientSecret, config)`.
     *
     * Applies a hard block for INR top-ups that would exceed [BALANCE_LIMIT_INR].
     */
    fun createTopUpIntent(amount: BigDecimal, currency: String = "INR") {
        // Hard block — INR limit
        if (currency == "INR" && wouldExceedLimit(amount)) {
            val current = _uiState.value
            if (current is WalletUiState.Success) {
                _uiState.value = current.copy(topUpResult = TopUpResult.FAILED)
            }
            return
        }

        viewModelScope.launch {
            _isTopUpLoading.value = true
            try {
                val response = api.createTopUpIntent(TopUpIntentRequest(amount, currency))
                pendingPaymentIntentId = response.paymentIntentId
                _clientSecret.value   = response.clientSecret
            } catch (e: Exception) {
                val prev = _uiState.value
                _uiState.value = WalletUiState.Error("Could not initiate top-up: ${e.message}")
                delay(2_500)
                _uiState.value = if (prev is WalletUiState.Success) prev
                                 else WalletUiState.Success(emptyList())
            } finally {
                _isTopUpLoading.value = false
            }
        }
    }

    // ── Step 2: confirm after Stripe SDK succeeds ─────────────────────────────

    fun confirmTopUp(paymentIntentId: String) {
        viewModelScope.launch {
            _isTopUpLoading.value = true
            _clientSecret.value   = null
            try {
                api.confirmTopUp(TopUpConfirmRequest(paymentIntentId))
                pendingPaymentIntentId = null
                loadAccounts()  // refresh balance display
                val current = _uiState.value
                if (current is WalletUiState.Success) {
                    _uiState.value = current.copy(topUpResult = TopUpResult.SUCCESS)
                }
            } catch (e: Exception) {
                val current = _uiState.value
                if (current is WalletUiState.Success) {
                    _uiState.value = current.copy(topUpResult = TopUpResult.FAILED)
                }
            } finally {
                _isTopUpLoading.value = false
            }
        }
    }

    // ── Called by WalletScreen when PaymentSheet reports failure ──────────────

    fun onPaymentSheetFailed(errorMessage: String?) {
        _clientSecret.value    = null
        pendingPaymentIntentId = null
        val current = _uiState.value
        if (current is WalletUiState.Success) {
            _uiState.value = current.copy(topUpResult = TopUpResult.FAILED)
        } else {
            _uiState.value = WalletUiState.Error(errorMessage ?: "Payment failed")
        }
    }

    fun clearTopUpResult() {
        _clientSecret.value = null
        val current = _uiState.value
        if (current is WalletUiState.Success) {
            _uiState.value = current.copy(topUpResult = null)
        }
    }
}

// ── Extensions ────────────────────────────────────────────────────────────────

private fun BigDecimal.coerceAtLeast(min: BigDecimal): BigDecimal =
    if (this < min) min else this

File: apps/ios/AegisPay/Features/Wallet/WalletViewModel.swift
import Foundation
import Combine

// ── Constants ─────────────────────────────────────────────────────────────────

/// Hard INR balance ceiling — Stripe KYC threshold.
/// Topping up is blocked when the INR account would exceed this.
let BALANCE_LIMIT_INR: Decimal = 100_000

// ── State ─────────────────────────────────────────────────────────────────────

enum TopUpPhase: Equatable {
    case idle
    case creatingIntent
    /// Stripe SDK should be presented with this clientSecret
    case awaitingStripeConfirmation(clientSecret: String)
    case confirmingWithBackend
    case success
    case failure(message: String)
}

// ── ViewModel ─────────────────────────────────────────────────────────────────

@MainActor
final class WalletViewModel: ObservableObject {

    // ── Published ──────────────────────────────────────────────────────────────
    @Published var accounts:     [Account]  = []
    @Published var isLoading:    Bool       = false
    @Published var errorMessage: String?    = nil
    @Published var topUpPhase:   TopUpPhase = .idle

    /// Live ECB-sourced FX rates (1 INR → X foreign currency).
    @Published private(set) var fxRates: FxRates = FxRates(usd: 0.01190, eur: 0.01099, gbp: 0.00936)
    @Published private(set) var fxLoading: Bool = false

    /**
     * The PaymentIntent ID (e.g. "pi_xxx") extracted from the clientSecret.
     * The WalletView passes this to `confirmTopUp` after Stripe SDK reports `.completed`.
     */
    private(set) var pendingPaymentIntentId: String? = nil

    // ── Dependencies ──────────────────────────────────────────────────────────
    private let walletService: WalletService

    init(walletService: WalletService = WalletService()) {
        self.walletService = walletService
    }

    // ── Load accounts + FX rates ──────────────────────────────────────────────

    func loadAccounts() {
        Task {
            isLoading    = true
            errorMessage = nil
            defer { isLoading = false }
            do {
                accounts = try await walletService.getMyAccounts()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
        loadFxRates()
    }

    func loadFxRates() {
        Task {
            fxLoading = true
            fxRates   = await FxRateService.shared.rates()
            fxLoading = false
        }
    }

    // ── Balance limit helpers ─────────────────────────────────────────────────

    /// INR available balance (nil when no INR account).
    var inrAvailable: Decimal? {
        accounts.first(where: { $0.currency == "INR" })?.availableBalance
    }

    /// Headroom left until the INR wallet cap.
    var inrRoom: Decimal {
        max(BALANCE_LIMIT_INR - (inrAvailable ?? 0), 0)
    }

    /// `true` when adding `amount` INR would breach the limit.
    func wouldExceedLimit(amount: Decimal) -> Bool {
        (inrAvailable ?? 0) + amount > BALANCE_LIMIT_INR
    }

    /// Human-readable wallet limit line, e.g. "Wallet limit: ₹1,00,000 (~£936 live)"
    func limitLine(currency: String) -> String {
        let limitFormatted = BALANCE_LIMIT_INR.formatted(currency: "INR")
        if currency == "INR" {
            return "Wallet limit: \(limitFormatted)"
        }
        if let converted = fxRates.convert(inr: BALANCE_LIMIT_INR, to: currency) {
            let convFormatted = converted.formatted(currency: currency)
            return "Wallet limit: \(limitFormatted) (~\(convFormatted) live)"
        }
        return "Wallet limit: \(limitFormatted)"
    }

    /// Headroom line, e.g. "Room: ₹99,800 (~£934 live)"
    func roomLine(currency: String) -> String {
        let room = inrRoom
        let roomFormatted = room.formatted(currency: "INR")
        if currency == "INR" {
            return "Room: \(roomFormatted)"
        }
        if let converted = fxRates.convert(inr: room, to: currency) {
            let convFormatted = converted.formatted(currency: currency)
            return "Room: \(roomFormatted) (~\(convFormatted) live)"
        }
        return "Room: \(roomFormatted)"
    }

    // ── Step 1: create PaymentIntent ──────────────────────────────────────────

    /**
     * Calls the backend to create a Stripe PaymentIntent.
     * On success transitions `topUpPhase` to `.awaitingStripeConfirmation(clientSecret:)`
     * so `WalletView` can build and present the Stripe `PaymentSheet`.
     */
    func startTopUp(amount: Decimal, currency: String = "INR") {
        guard topUpPhase == .idle else { return }
        topUpPhase = .creatingIntent
        Task {
            do {
                let intent = try await walletService.createTopUpIntent(amount: amount, currency: currency)
                // Store the pi_xxx so confirmTopUp can send it to the backend
                pendingPaymentIntentId = intent.paymentIntentId
                topUpPhase = .awaitingStripeConfirmation(clientSecret: intent.clientSecret)
            } catch {
                topUpPhase = .failure(message: "Could not start top-up: \(error.localizedDescription)")
            }
        }
    }

    // ── Step 2: confirm after Stripe SDK succeeds ─────────────────────────────

    /**
     * Called by `WalletView` when `PaymentSheetResult.completed` fires.
     * Notifies the backend to credit the user's ledger balance.
     */
    func confirmTopUp(paymentIntentId: String) {
        topUpPhase = .confirmingWithBackend
        Task {
            do {
                try await walletService.confirmTopUp(paymentIntentId: paymentIntentId)
                pendingPaymentIntentId = nil
                topUpPhase = .success
                loadAccounts()   // refresh balance
            } catch {
                topUpPhase = .failure(
                    message: "Payment confirmed but failed to credit wallet: \(error.localizedDescription)"
                )
            }
        }
    }

    // ── Called by WalletView on PaymentSheet failure ───────────────────────────

    func handlePaymentSheetFailure(_ message: String) {
        pendingPaymentIntentId = nil
        topUpPhase = .failure(message: message)
    }

    func resetTopUpPhase() {
        pendingPaymentIntentId = nil
        topUpPhase = .idle
    }
}
